### DISCUSSION TITLE
Backrest on BTRFS snapshots, email diff

### DISCUSSION THREAD

**Discussion by lucmos** [garethgeorge/backrest#320] [Q&A] [answered]
Hello, 
I would like to abandon my hacky scripts to perform backups with restic  :smile: 
After searching for a while, I think backrest would suit my usecase perfectly!

My scripts work in the following way:

```bash
#!/usr/bin/bash

btrfs subvolume snapshot -r /srv/dev-disk-by-label-data/luca/ /srv/dev-disk-by-label-data/.snapshots/srv/dev-disk-by-label-data/luca

echo -e "\n\n~~~~~\n\n"

. /root/restic_envs_passport.sh
echo -e "\n\n\n---\nBACKUP TO: ${RESTIC_REPOSITORY}\n---\n"
/root/bin/restic --verbose backup "/srv/dev-disk-by-label-data/.snapshots/srv/dev-disk-by-label-data" --tag data --exclude-file="/root/backup-exclude.txt"
echo -e "\n---\nDIFF\n---\n"
/root/bin/restic snapshots --json | python -c "import sys, json; j=json.load(sys.stdin); j = [x for x in j if 'tags' not in x or 'fireflyiii' not in x['tags'] or 'actual' not in x['tags']]; print(j[-2]['short_id'] + ' ' + j[-1]['short_id'])" | xargs /root/bin/restic diff

echo -e "\n\n~~~~~\n\n"

btrfs subvolume delete /srv/dev-disk-by-label-data/.snapshots/srv/dev-disk-by-label-data/luca
```

The idea is the following:

1. The backup scripts first creates btrfs snapshots, to freeze the state and avoid inconsistent backups (in case the state changes while the backup is in progress. A classic example is backing up a live database, that often is unreadable without this kind of state freeze)
2. Execute restic backup
3. Print the differences with an hacky one liner. This is useful because, since it's executed in a crontab, these differences are emailed directly to me
4. Remove the btrfs temporary snapshots

I am interested in automating all this with backrest. 
Could you give some pointer on what is the best way to do it?

Any help is much appreciated! :smile:

**Comment by garethgeorge** [comment]
Hmm, this is pretty challenging in backrest today -- backrest doesn't support a dynamic path for backup and it looks like restic won't follow symlinks.

You'll need a way to mount the btrfs snapshot in a consistent location accessible to backrest. In terms of actually running scripts before/after backup that should be doable with the CONDITION_BACKUP_START and CONDITION_BACKUP_END set on a command hook.

**Reply by Hoempi** [reply]
I actually found this as I'm plannning on doing basically the same. With one key difference: My Btrfs snapshot gets mounted on the same path everytime. So that should be a doable in general, as far as I understand.

My question, being pretty new to using Btrfs snapshots like this, is whether this could be done when running the Backrest Docker container. Probably not, as that would result in the Docker container needing a great level of access to the filesystem of the host system, correct?

So, I assume I would need to install Backrest directly on the host system to reach this.

**Reply by Hoempi** [reply, marked as answer]
I just went ahead and tested this. Works like a charm, this is my main backup plan:
```
{
  "id": "main",
  "repo": "optiplex",
  "paths": [
    "/mnt/docker-data/docker-data-snapshot",
    "/opt/stacks"
  ],
  "excludes": [],
  "iexcludes": [],
  "schedule": {
    "clock": "CLOCK_LOCAL",
    "cron": "0,42 * * * *"
  },
  "retention": {
    "policyTimeBucketed": {
      "yearly": 1,
      "monthly": 3,
      "weekly": 4,
      "daily": 7,
      "hourly": 5
    }
  },
  "hooks": [
    {
      "conditions": [
        "CONDITION_SNAPSHOT_START",
        "CONDITION_SNAPSHOT_ERROR",
        "CONDITION_SNAPSHOT_SUCCESS"
      ],
      "actionHealthchecks": {
        "webhookUrl": "https://myhc.example.com/ping/76000475-this-is-a-dummy"
      }
    },
    {
      "conditions": [
        "CONDITION_SNAPSHOT_START"
      ],
      "onError": "ON_ERROR_FATAL",
      "actionCommand": {
        "command": "btrfs subvolume snapshot -r /mnt/docker-data /mnt/docker-data/docker-data-snapshot"
      }
    },
    {
      "conditions": [
        "CONDITION_SNAPSHOT_END"
      ],
      "onError": "ON_ERROR_CANCEL",
      "actionCommand": {
        "command": "btrfs subvolume delete /mnt/docker-data/docker-data-snapshot"
      }
    }
  ]
}
```

I also really like that the healtcheck hooks automatically choose the correct path, e.g. `https://myhc.example.com/ping/76000475-this-is-a-dummy/start`

**Reply by garethgeorge** [reply]
Really neat that this works with your btrfs setup today and very cool to see scripting used this way. If you don't mind, I might actually steal this as an example for the [cookbooks section on the wiki](https://garethgeorge.github.io/backrest/cookbooks/command-hook-examples). Is it still working well for you?

**Reply by Hoempi** [reply]
Yeah, it's still working well, and sure, feel free. Glad if it may help someone else tackling a problem.

**Reply by robflate** [reply]
@Hoempi Are you saying this is working using the Backrest Docker image? Or is this a bare metal installation of Backrest. Thanks.

**Reply by Hoempi** [reply]
I switched to baremetal, yes. Else it would get complicated to run the Btrfs snapshot commands on the host.

**Reply by robflate** [reply]
Thanks. I really like this concept. I have some general questions about using this technique;

My Backrest Plans have lots of Includes and Excludes. When adding these, Backrest auto fills the path when you start typing in the `Excludes:` text box;

![Screenshot from 2024-12-02 21-44-43](https://github.com/user-attachments/assets/6a703f71-4bde-4d5d-a00e-be658b3df794)

This won't work because the snapshot won't actually exist until the `CONDITION_SNAPSHOT_START` Hook runs the `btrfs subvolume snapshot ...` command.

Solutions;

- I could do all my Includes/Excludes on my actual folders (e.g. `/mnt/docker-data/example`) and then when I'm done, change all the paths (e.g. `/mnt/docker-data/docker-data-snapshot/example`).
- I could run a btrfs snapshot before making/editing my Backrest Plan so that `/mnt/docker-data/docker-data-snapshot/example` would actually exist and the auto fill would work. I'd then have to delete the snapshot otherwise the btrfs command will return an error indicating that the snapshot already exists when the `CONDITION_SNAPSHOT_START` Hook runs.
- A better solution??

Any help appreciated. Thanks.

**Reply by robflate** [reply]
A quick follow up question for @Hoempi. Did you set the `USER_SUBVOL_RM_ALLOWED` mount option in your `/etc/fstab` file to be able to run `btrfs subvolume delete` withour sudo?

Edit: Just noticed you're creating the snapshot as read-only so you'd need elevated privileges to delete it. Are you running as root user?

**Reply by Hoempi** [reply]
Hey @robflate,

as for your first question I would probably go with manually taking a snapshot before configuring and deleting it afterwards. I personally strongly prefer using auto-completion, as I sometimes just have stubby fingers.

As for your second question I would have to check when I get back home. But if I'm not mistaken backrest is running as a systemd service and thus has the required privileges for taking and deleting snapshots implicitly.

**Reply by Hoempi** [reply]
> Did you set the `USER_SUBVOL_RM_ALLOWED` mount option in your `/etc/fstab` file to be able to run `btrfs subvolume delete` withour sudo?

Hey @robflate ,

I just got back to my desk at home and checked. I do not have `USER_SUBVOL_RM_ALLOWED` in `/etc/fstab`. So, I guess I'm currently running my systemd service for backrest with root privileges. Guess, I should check on this. Thanks for pointing it out!
